MM-16730: e2e test for replying to an older message

matthewshirley opened the pull request:
### Summary
Introduce an e2e test case for replying to an older message as identified in MM-16730. 🤖

#### Ticket Link
GitHub Issue: Fixes https://github.com/mattermost/mattermost-server/issues/11570
Jira ticket: https://mattermost.atlassian.net/browse/MM-16730

#### Related Pull Requests
N/A

#### Notes
- I have modified the postMessageAs function to return the status, data and post ID. Existing functions (getLastPostId and getNthPostId) would introduce a fragile test.

I would appreciate/love feedback on this PR. 💥

thekiiingbob commented on e2e/cypress/integration/channel/message_spec.js:
```diff
@@ -140,4 +141,63 @@ describe('Message', () => {
+    it('MM-16730 Reply to an older message', () => {
+        cy.getCurrentChannelId().then((channelId) => {
+            // # Get yesterdays date in UTC
+            const yesterdaysDate = new Date().getTime() - 86400000;
```
Nit pick, but could leverage moment (included in Cypress) for possibly easier "historic" dates.

matthewshirley replied:
Hooray for moment 🎉 Updated in e84aae0.

thekiiingbob commented on e2e/cypress/integration/channel/message_spec.js:
```diff
@@ -140,4 +141,63 @@ describe('Message', () => {
+            cy.getLastPostId().then((replyId) => {
+                // * Verify that the reply is in the channel view with matching text
+                cy.get(`#post_${replyId}`).within(() => {
+                    cy.get('.post__link').should('be.visible').and('have.text', 'Commented on sysadmin\'s message: Hello from yesterday');
```
Haven't looked at the react code yet, but could you update the `.post__link` element to use an id?

matthewshirley replied:
I will add an ID in for that element. Is there a preferred format? "#postLink" or "#postLink_{childPost.id}".

thekiiingbob replied:
`#postLink` (like you have done) should be fine here. Especially since you are doing it inside a `within`. If we were to want to get the post link without the `within`, I'd say it should be a `data-testid` attribute with the id of the post slapped on it as well. No changes required IMO at the moment though.

matthewshirley replied:
Cool - I was in there for the latter feedback so adjusted this.

thekiiingbob commented on e2e/cypress/integration/channel/message_spec.js:
```diff
@@ -140,4 +141,63 @@ describe('Message', () => {
+
+        cy.get('@yesterdaysPost').then((postId) => {
+            // # Open RHS comment menu
+            cy.clickPostCommentIcon(postId);
```
Depending on the volume of messages, this step fails for me because "yesterdays" post is pushed up beyond where it's visible. Tacking on some `scrollIntoViews()` where appropriate should address this.

matthewshirley replied:
Nice catch. I'll add that in.

My understanding is that it's only needed before the `clickPostCommentIcon` since there are no actions on the other posts in the center screen. Everything should remain in the DOM.

Let me know if my understanding is specious.

thekiiingbob replied:
Eventually, we will be parallelizing these tests so we want to guard against other tests possibly writing to this same channel, what you did seems perfectly fine.

thekiiingbob commented on e2e/cypress/support/task_commands.js:
```diff
@@ -11,7 +11,12 @@
+        expect(status).to.equal(201);
+
+        // # Return the data so it can be interacted in a test
+        cy.wrap({id: data.id, status, data});
```
Good call. Nice to have!

srkgupta commented on e2e/cypress/integration/channel/message_spec.js:
```diff
@@ -140,4 +141,66 @@ describe('Message', () => {
+                cy.get(`#rhsPost_${replyId}`).within(() => {
+                    cy.get('#postLink').should('not.be.visible');
+                    cy.get(`#postMessageText_${replyId}`).should('be.visible').and('have.text', 'A reply to an older post with attachment');
+                    cy.get('.post-image__column').should('be.visible').and('have.text', 'mattermost-icon.png');
```
Currently the testcase fails here, because the element `.post-image__column` is not found. I guess you are referring to the element `.image-name`. Can you please recheck and update the locator correctly. 
![image](https://user-images.githubusercontent.com/1429138/62287133-df7a8500-b476-11e9-94a8-21ec189dca48.png)

thekiiingbob replied:
Same deal here with the prefernece to use an id for the `.post_image__column`. Depending on the react code, of course.

matthewshirley replied:
Interesting, the file displays differently for me.

<img width="271" alt="Screen Shot 2019-08-01 at 9 57 25 am" src="https://user-images.githubusercontent.com/450704/62312392-e7b8cd00-b442-11e9-92ec-d5084d171584.png">

`.post-image__column` is visible however `.image-name` does not. I'll look into that and add a ID.

matthewshirley replied:
Fixed in 3eba67f.

thekiiingbob commented on e2e/cypress/integration/channel/message_spec.js:
```diff
@@ -140,4 +141,66 @@ describe('Message', () => {
+    it('MM-16730 Reply to an older message', () => {
+        cy.getCurrentChannelId().then((channelId) => {
+            // # Get yesterdays date in UTC
+            const yesterdaysDate = Cypress.moment().subtract(1, 'days').valueOf();
```
👍

thekiiingbob commented:
@matthewshirley I think we need to make some modifications to this test a bit to make it a bit safer. We talked about this a bit [here](https://community-daily.mattermost.com/core/pl/gkhiq87k3tbm3ycx5h5ppza8ua) 

Effectively, say you run this test a handful of times in a row, it's possible for the "yesterday" post to live above the scroll point where you would need to load more posts.

To address this, I think you should create a new channel at the beginning of the test so that there is no history that you need to account for, everything else should be able to remain as is.

Let me know if you have any questions!

matthewshirley commented:
> @matthewshirley I think we need to make some modifications to this test a bit to make it a bit safer. We talked about this a bit [here](https://community-daily.mattermost.com/core/pl/gkhiq87k3tbm3ycx5h5ppza8ua)
> 
> Effectively, say you run this test a handful of times in a row, it's possible for the "yesterday" post to live above the scroll point where you would need to load more posts.
> 
> To address this, I think you should create a new channel at the beginning of the test so that there is no history that you need to account for, everything else should be able to remain as is.
> 
> Let me know if you have any questions!
Fixed in ac55277.
 
Safe is fun. I've updated the test to create and destroy a channel respectively. I also separated the test from the message_spec since I noticed other PRs could fit into that scope (https://github.com/mattermost/mattermost-server/issues/11574).

Let me know your thoughts.
